path_roles.go

```diff
@@ -320,14 +357,136 @@ func (b *azureSecretBackend) pathRoleList(ctx context.Context, req *logical.Requ
 func (b *azureSecretBackend) pathRoleDelete(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
 	name := d.Get("name").(string)
 
-	err := req.Storage.Delete(ctx, fmt.Sprintf("%s/%s", rolesStoragePath, name))
+	lock := locksutil.LockForKey(b.appLocks, name)
+	lock.Lock()
+	defer lock.Unlock()
+
+	role, err := getRole(ctx, name, req.Storage)
+	if err != nil {
+		return nil, errwrap.Wrapf(fmt.Sprintf("unable to get role %s: {{err}}", name), err)
+	}
+	if role == nil {
+		return nil, nil
+	}
+
+	err = req.Storage.Delete(ctx, fmt.Sprintf("%s/%s", rolesStoragePath, name))
 	if err != nil {
 		return nil, errwrap.Wrapf("error deleting role: {{err}}", err)
 	}
 
+	if role.ApplicationType == "dynamic" {
+		_, err = b.deleteApp(ctx, req, role)
+		if err != nil {
+			return &logical.Response{Warnings: []string{"error deleting dynamic Azure service principal"}}, nil
+		}
+	} else if role.ApplicationType == "static" {
+		c, err := b.getClient(ctx, req.Storage)
+		if err != nil {
+			return nil, errwrap.Wrapf("error during revoke: {{err}}", err)
+		}
+		for _, creds := range role.Credentials {
+			err = c.deleteAppPassword(ctx, role.ApplicationObjectID, creds.KeyId)
```

Comment:
Should this delete all credentials created via the plugin, not just the last?

Reply:
Yeah, so this is the case where the Vault role is deleted. But if an existing SP is used, we need to clean up all password credentials that were created for that SP which is why we delete all credentials that are tracked.